lib/temporal/worker/reactor/promise.rb

```diff
@@ -0,0 +1,57 @@
+module Temporal
+  class Worker
+    class Reactor
+      class Promise
+        class AlreadyFulfilledError < StandardError; end
+
+        def initialize
+          @result = nil
+          @fulfilled = false
+        end
+
+        def fulfilled?
+          @fulfilled
+        end
+
+        def result
+          wait
+
+          raise @exception if defined?(@exception)
+
+          @result
+        end
+
+        def wait
+          return if fulfilled?
+
+          @fiber = Fiber.current
+
+          Fiber.yield
+        end
+
+        def set(value)
```

Comment:
Would recommend an explicit `set_exception` instead of using exception type to distinguish